### ISSUE TITLE
Publish vector tiles for download

### ISSUE THREAD

**Issue by sfkeller** [osm2vectortiles/osm2vectortiles#68]
I already talked to IT staff at HSR and they are helping us. I apply for one VM and 100 GB storage.
 
According to Petr there will be a 30 to 50 GB planet file and I suggest that in addition this is cut into about 200 sovereign states (countries), which sums up to another 50 GB.

Questions:
 * Is this storage estimation enough?
 * What amount of traffic do we estimate? Around 100 planet downloads per day (= 5 TB)?
 * What could be a good name: "Open Streets Vector Tiles"?

**Comment by lukasmartinelli** [comment]
> I already talked to IT staff at HSR and they are helping us. I apply for one VM and 100 GB storage.

Thanks alot - that is perfect!

> Is this storage estimation enough?

Should be enough - more is better of course but the most important thing is that we can host the planet file on a network where we don't have to pay much for the traffic.

> What amount of traffic do we estimate? Around 100 planet downloads per day (= 5 TB)?

That's the best case. We won't get over 10 downloads a day in the start phase - and if Switch (but they have so much excess capacity) cannot handle the load we could still switch to a CDN.

> What could be a good name: "Open Streets Vector Tiles"?

All we need is a simple mirror to download files. Only needs to be an HTTP server like Apache which can serve large files for download. We can manage the download links from http://osm2vectortiles.org/data/download.html and just point to the server download urls.

**Comment by sfkeller** [comment]
Ok. You say IP address are enough for serving these files.
But in order to "sell" this to the HSR-IT I still need a I still need a project name.
 * I've seen you call it "OSM Vector Tiles",
 * What about "Open Streets Vector Tiles"?

**Comment by klokan** [comment]
Great news on the academic Switch download hosting possibility...

I guess suitable project name for HSR-IT would be: osm2vectortiles

On Saturday, 7 November 2015, Stefan <notifications@github.com> wrote:

> Ok. You say IP address are enough for serving these files.
> But in order to "sell" this to the HSR-IT I still need a I still need a
> project name.
>
>    - I've seen you call it "OSM Vector Tiles",
>    - What about "Open Streets Vector Tiles"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geometalab/osm2vectortiles/issues/68#issuecomment-154751827>
> .
>


-- 
Petr Pridal, Ph.D.
CEO

Klokan Technologies GmbH
Hofnerstrasse 98, 6314 Unterageri, Switzerland
Tel: +41 (0)41 511 26 12
Email: info@klokantech.com
Web: http://www.klokantech.com/

**Comment by sfkeller** [comment]
Little bit off topic: But 
1. how would point clustering fit into vector tiles format?
2. and is there a fast query to count the total no. of features in a tile?

**Comment by lukasmartinelli** [comment]
> 1. how would point clustering fit into vector tiles format?

One approach to display millions and billions of points for example is using https://github.com/mapbox/tippecanoe

This actually works beautifully, one does pipe in GeoJSON features and tippecanoe will automatically drop points as it moves to lower level zoom layers.

That's how I was able to display 10 million points http://lukasmartinelli.ch/swiss-location/
I don't know how well Mapbox GL handles that many features but there is a recommended max feature count per layer additional to the size limit of 500KB.

> 2. and is there a fast query to count the total no. of features in a tile?

The compare script @manuelroth wrote is able to do that.
It parses the vector tile format, takes a look into each layer and counts the features (among other things).

**Comment by klokan** [comment]
@sfkeller could you please report what is the status of the hosting possibility on HSR?

**Comment by sfkeller** [comment]
It was stalled around weeks but last week the responsable IT support guy got green light to investigate further. I think he has already a test account for us. I'll be back again soon...

**Comment by klokan** [comment]
SWITCHengines is the Amazon AWS for Swiss Academia (powered by OpenStack) - this would be great to have access to indeed - especially for the EC2 equivalent (compute machines). If we get there a virtual machine it is great...

For the "download hosting" the simplest would be probably just a public bucket in SWITCHEngines for storing ~100-150 GB data:  https://help.switch.ch/engines/documentation/s3-like-object-storage/ (which is OpenStack Swift I guess).

**Comment by manuelroth** [comment]
I didn't know that Switch also provides Cloud Services. Would be really nice, if we could have a public bucket.

**Comment by manuelroth** [comment]
I tried to create a bucket and upload an mbtiles to SWITCHengines S3 Object Store. The upload worked fine, but somehow I can't access the mbtiles file. In AWS S3 a bucket policy needs to be created to allow access to a bucket. But there is no word on that in the [documentation](https://help.switch.ch/engines/documentation/s3-like-object-storage/). Because of that I created these two issues on their forum([issue1](https://forum.cloud.switch.ch/t/s3-object-storage-bucket-url-schema/38), [issue2](https://forum.cloud.switch.ch/t/s3-object-storage-bucket-policy/39))

**Comment by manuelroth** [comment]
The support-team of SWITCHengines told me that they do not support bucket policies only acl's. I used the [s3cmd tool](http://s3tools.org/s3cmd) to create a bucket and make it publicly accessible. 

This command creates a bucket osm2vectortiles and makes it and everything inside the bucket public:
```bash
s3cmd mb s3://osm2vectortiles/ --acl-public --recursive
```
The bucket can now be accessed from this url: https://osm2vectortiles.os.zhdk.cloud.switch.ch/

**Comment by klokan** [comment]
Great - this sounds great. I can get the file download https://osm2vectortiles.os.zhdk.cloud.switch.ch/Wolken.jpg directly.

Is CNAME supported (so pointing downloads.osm2vectortiles.org to osm2vectortiles.os.zhdk.cloud.switch.ch) to use it instead?

I am really keen to get my Switchengines account as well, if possible - so that I can see their system...

**Comment by klokan** [comment]
Aha, it seems you have just deleted the Wolken.jpg file ;-)

**Comment by manuelroth** [comment]
:smile: exactly. I want to try out [mapbox tile copy](https://github.com/mapbox/mapbox-tile-copy), it takes a mbtiles, uploads it to S3 and extracts it to the format /{z}/{x}/{y}.

```bash
mapbox-tile-copy ~/data/my-tiles.mbtiles s3://my-bucket/folder/mbtiles/{z}/{x}/{y}
```
Then we could serve the vector tiles statically, like mapbox does it with mapbox streets.

**Comment by klokan** [comment]
I don't see a real reason to make this test. The bucket is required for downloading individual large .mbtiles.

Technically hosting of vector tiles this way is possible - but it makes not much sense for the world - (upload would take an year probably) - for Zurich or smaller are you can do that of course, but you can do that on github too. The hosting should support CORS. See: countries folder in https://github.com/klokantech/mapbox-gl-js-offline-example ...

**Comment by manuelroth** [comment]
I just think it would be nice, if we had an alternative to downloading the extract mbtiles. We could provide a Mapbox GL example which contains an url to the vector tiles. This would be way quicker than downloading the mbtiles file and starting our tile server container. It would be nice, if we could provide something for the people, which just want to have a quick look on our map. 

I uploaded the mbtiles file of switzerland to the bucket osm2vectortiles. It can be accessed on this link: https://osm2vectortiles.os.zhdk.cloud.switch.ch/switzerland.mbtiles

**Comment by manuelroth** [comment]
Gist with BBOXs of alle countries of the world.

**Comment by lukasmartinelli** [comment]
Upload to Google Drive with the https://github.com/Grive/grive failed several times even with max. part size. This made me wonder - when users will download such a big file like 54 GB over HTTP they will nearly always run into download problems.

With the `wget -c` continue option it should be better but  downloading files without tools for special resumable HTTP downloads is not really easy for people who don't want to use the command line.

I suggest we serve the big MBTiles file via Torrent as well (we can run it on the SwitchCloud VM).

Usually people split up files like these into several parts that need to be merged again (and then use tools like e.g. http://jdownloader.org/). For the GitHub archive they do it like this https://www.githubarchive.org/. But then again for OSM planet they also provide a 46 GB planet file via HTTP - so it definitly works as well.

**Comment by lukasmartinelli** [comment]
File available for download: http://downloads.osm2vectortiles.org/world.mbtiles

A vector tile server is set up as well using tileserver-php: http://vectortiles.osm2vectortiles.org/
In Mapbox Studio Classic a user can now simply switch the TileJSON URL to http://vectortiles.osm2vectortiles.org/world.json and can start designing.

**Comment by lukasmartinelli** [comment]
Basic and OSM Bright are now available.
Sadly all other available Mapbox Studio themes also use Mapbox Terrain.

- http://osm2vectortiles.org/maps/basic-v8.html
- http://osm2vectortiles.org/maps/bright-v8.html

The map is now embedded directly on the start page. Not sure if good idea yet.

![screenshot from 2015-12-04 20-54-36](https://cloud.githubusercontent.com/assets/1288339/11599833/61e5d222-9ac9-11e5-99a5-3bf4e1c82b1a.png)

**Comment by klokan** [comment]
All of the above are perfect news!

Vector tiles uploaded to Drive for backup - it took less then 30 minutes from our server!
@lukasmartinelli I have recommended you `rclone` utility for working with Drive in the past.

**Comment by lukasmartinelli** [comment]
I also set up a tessera based rastertile server for all mapbox-studio classic projects that use mapbox streets. And overview of the all the supported maps can be found at: http://osm2vectortiles.org/maps/

The server is also running on Switch http://rastertiles.osm2vectortiles.org on a different smaller server.


![screenshot from 2015-12-10 13-54-15](https://cloud.githubusercontent.com/assets/1288339/11715829/8b214740-9f45-11e5-85bf-86a56f2fb298.png)

I also found some severe bugs in this process :cry: , like too long labels in the carribean and a tile loading problem for new york (as if the tiles there are missing!).
